src/imitation/algorithms/bc.py

```diff
@@ -4,178 +4,261 @@
+                        f"{epoch_num} -- did it reset correctly?",
+                    )
+                if self.on_epoch_end is not None:
+                    self.on_epoch_end(epoch_num)
+
+        # Note: the islice here ensures we do not exceed self.n_batches
+        return itertools.islice(batch_iterator(), self.n_batches)
+
+
+@dataclasses.dataclass(frozen=True)
+class BehaviorCloningLoss:
+    """Container for the different components of behavior cloning loss."""
+
+    neglogp: th.Tensor
+    entropy: th.Tensor
+    ent_loss: th.Tensor
+    prob_true_act: th.Tensor
+    l2_norm: th.Tensor
+    l2_loss: th.Tensor
+    loss: th.Tensor
+
+
+@dataclasses.dataclass(frozen=True)
+class BehaviorCloningLossCalculator:
+    """Functor to compute the loss used in Behavior Cloning."""
+
+    ent_weight: float
+    l2_weight: float
+
+    def __call__(
+        self,
+        policy: policies.ActorCriticPolicy,
+        obs: Union[th.Tensor, np.ndarray],
+        acts: Union[th.Tensor, np.ndarray],
+    ) -> BehaviorCloningLoss:
+        """Calculate the supervised learning loss used to train the behavioral clone.
 
         Args:
-            lr: the constant learning rate that calls to this object will return.
+            policy: The actor-critic policy of which to compute the loss.
+            obs: The observations seen by the expert.
+            acts: The actions taken by the expert.
+
+        Returns:
+            A BehaviorCloningLoss object with the loss and all the components it
+            consists of.
         """
-        self.lr = lr
+        _, log_prob, entropy = policy.evaluate_actions(obs, acts)
+        prob_true_act = th.exp(log_prob).mean()
+        log_prob = log_prob.mean()
+        entropy = entropy.mean()
+
+        l2_norms = [th.sum(th.square(w)) for w in policy.parameters()]
+        l2_norm = sum(l2_norms) / 2  # divide by 2 to cancel with gradient of square
+
+        ent_loss = -self.ent_weight * entropy
+        neglogp = -log_prob
+        l2_loss = self.l2_weight * l2_norm
+        loss = neglogp + ent_loss + l2_loss
 
-    def __call__(self, _):
-        """Returns the constant learning rate."""
-        return self.lr
+        return BehaviorCloningLoss(
+            neglogp=neglogp,
+            entropy=entropy,
+            ent_loss=ent_loss,
+            prob_true_act=prob_true_act,
+            l2_norm=l2_norm,
+            l2_loss=l2_loss,
+            loss=loss,
+        )
 
 
-class _NoopTqdm:
-    """Dummy replacement for tqdm.tqdm() when we don't want a progress bar visible."""
+@dataclasses.dataclass(frozen=True)
+class BehaviorCloningTrainer:
+    """Functor to fit a policy to expert demonstration data."""
 
-    def close(self):
-        pass
+    loss: BehaviorCloningLossCalculator
+    optimizer: th.optim.Optimizer
+    policy: policies.ActorCriticPolicy
+    device: th.device  # TODO(max): not sure whether the device belongs in this class
```

Comment:
I think we can use `self.policy.device` instead; the only reason we need it for this class AFAICT is to send the inputs to `self.device` before evaluating the loss which is all in terms of `self.policy`.

Reply:
Thanks for the hint! Implemented it!